client/app/components/permissions-editor/PermissionsEditorDialog.jsx

```diff
@@ -0,0 +1,183 @@
+  const addPermission = useCallback((userId, accessType = 'modify') => $http.post(
+    url, { access_type: accessType, user_id: userId },
+  ).catch(() => notification.error('Could not grant permission to the user'), [url]));
+
+  const removePermission = useCallback((userId, accessType = 'modify') => $http.delete(
+    url, { data: { access_type: accessType, user_id: userId } },
+  ).catch(() => notification.error('Could not remove permission from the user')), [url]);
+
+  return { loadGrantees, addPermission, removePermission };
+}
+
+const searchUsers = searchTerm => User.query({ q: searchTerm }).$promise
+  .then(({ results }) => results)
+  .catch(() => []);
+
+function PermissionsEditorDialogHeader({ context }) {
+  return (
+    <>
+      Manage Permissions
+      <div className="modal-header-desc">
+        {`Editing this ${context} is enabled for the users in this list and for admins. `}
+        <HelpTrigger type="MANAGE_PERMISSIONS" />
+      </div>
+    </>
+  );
+}
+
+PermissionsEditorDialogHeader.propTypes = { context: PropTypes.oneOf(['query', 'dashboard']) };
+PermissionsEditorDialogHeader.defaultProps = { context: 'query' };
+
+function UserSelect({ onSelect, previewCardAddon, isUserDisabled }) {
+  const [loadingUsers, setLoadingUsers] = useState(true);
+  const [users, setUsers] = useState([]);
+  const [searchTerm, setSearchTerm] = useState('');
+
+  const debouncedSearchUsers = useCallback(debounce(
+    search => searchUsers(search)
+      .then(setUsers)
+      .finally(() => setLoadingUsers(false)),
+    DEBOUNCE_SEARCH_DURATION,
+  ), []);
+
+  useEffect(() => {
+    setLoadingUsers(true);
+    debouncedSearchUsers(searchTerm);
+  }, [searchTerm]);
+
+  return (
+    <Select
+      className="w-100 m-b-10"
+      placeholder="Add users..."
+      showSearch
+      onSearch={setSearchTerm}
+      suffixIcon={loadingUsers ? <i className="fa fa-spinner fa-pulse" /> : <i className="fa fa-search" />}
+      filterOption={false}
+      notFoundContent={null}
+      value={undefined}
+      getPopupContainer={trigger => trigger.parentNode}
+      onSelect={onSelect}
+    >
+      {users.map(user => (
+        <Option key={user.id} value={user.id} disabled={isUserDisabled(user)}>
+          <UserPreviewCard user={user}>
+            {previewCardAddon(user)}
+          </UserPreviewCard>
+        </Option>
+      ))}
+    </Select>
+  );
+}
+
+UserSelect.propTypes = {
+  onSelect: PropTypes.func,
+  previewCardAddon: PropTypes.func,
+  isUserDisabled: PropTypes.func,
+};
+UserSelect.defaultProps = { onSelect: () => {}, previewCardAddon: () => null, isUserDisabled: () => false };
+
+function PermissionsEditorDialog({ dialog, owner, context, aclUrl }) {
+  const [loadingGrantees, setLoadingGrantees] = useState(true);
+  const [grantees, setGrantees] = useState([]);
+  const { loadGrantees, addPermission, removePermission } = useGrantees(aclUrl);
+  const loadUsersWithPermissions = useCallback(() => {
+    setLoadingGrantees(true);
```

Comment:
Swapping the list for a loading indicator on each add and remove is quite jarring.
How about adding a small indicator to the top right of the scrollbox?

Reply:
That was bothering me too, I was waiting for you to say sth about it. I tried a few options like using an indicator in the Search bar, but ended up with more complexity and not satisfactory. Will try that one :+1: